fcrepo4/fcrepo4
# FCREPO-3399 - External URI case sensitivity and handling changes

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3399

# What does this Pull Request do?
* Backports fix for external uri validation case sensitivity issues with file uris from Fedora 6
* Backports most of the external binary integration tests from Fedora 6
* Fixes bug when switching between redirect and proxy handling on an existing external binary, where it would retain the previous content uri and handling in modeshape. This would result in the "proxy" handling taking precedence, even if it was not the type selected by more recent updates.

# How should this be tested?

Run automated tests. For notes on how to manually test, see:
https://github.com/fcrepo4/fcrepo4/pull/1735

If on a file system that is case sensitive (not macOs), can try creating an external binary with capital letters in it, which should succeed. Can then try switching it back and forth to proxy/redirect, the handling should reflect your updates.

```
mvn -Dfcrepo.home=target/fcrepo -Dfcrepo.external.content.allowed=fcrepo-http-api/src/test/resources/allowed_external_paths.txt clean jetty:run -pl fcrepo-webapp -Denforcer.skip -Dmaven.javadoc.skip=true


# create external binary with a file containing capital letters, with proxy handling
curl -i -H"Link: <file:///Users/bbpennel/git/fcrepo4/README.md>; rel=\"http://fedora.info/definitions/fcrepo#ExternalContent\"; handling=\"proxy\"; type=\"text/plain\"" -XPUT -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test_case_sensitive

# Head request should indicate it would return the file directly
curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test_case_sensitive

HTTP/1.1 200 OK
Content-Location: file:/Users/bbpennel/git/fcrepo4/README.md


# Replace the binary with a different one, using redirect
curl -i -H"Link: <file:///Users/bbpennel/git/fcrepo4/pom.xml>; rel=\"http://fedora.info/definitions/fcrepo#ExternalContent\"; handling=\"redirect\"; type=\"text/plain\"" -XPUT -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test_case_sensitive

# Head request should indicate that it would redirect to the file
curl -I -ufedoraAdmin:fedoraAdmin http://localhost:8080/rest/test_case_sensitive

HTTP/1.1 307 Temporary Redirect
Location: file:/Users/bbpennel/git/fcrepo4/pom.xml
```
